feat: added permissions (pull-requests: read) setting to workflow and README example

natterstefan opened the pull request:
## Changes

I see more and more actions taking care of managing the permissions of the `GITHUB_TOKEN,` which is why I added it to your example and the workflows (+ updated them to the latest version).

The ones I added to `release,` and `versioning` might need some changes once you merge them. I'm pretty sure that's all they need.

## Docs

- [Assigning permissions to jobs - GitHub Docs](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)
-  [GitHub Actions: Control permissions for GITHUB_TOKEN | GitHub Changelog](https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/)
- [Workflow syntax for GitHub Actions - GitHub Docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions)

amannn commented:
Hey @natterstefan,

it's nice to see you here! 🙂 Thank you for your PR!

There was some discussion on this in #182 (likely also #212 intended to be about this). I was previously a bit opposed to this since it requires more configuration, but maybe it can in fact be helpful for trading that with trustworthiness.

Do you have examples of other actions which advertise the usage of permissions?

Some considerations (unfortunately GitHub freezes today for me when I open the files tab, so I'll list the feedback here):
- [ ] When the `wip` feature is used, we also require to update statuses. Can you add information about this in the comment for the `wip` option? I think generally suggesting read permissions for PRs is a good idea.
- [ ] Can you please also link to a test run as it's mentioned in the [contributors guide](https://github.com/amannn/action-semantic-pull-request/blob/main/CONTRIBUTORS.md)? E.g. I'd be curious if using the checkout action only requires the `pull-requests: read` option, or if a permission like `contents` is needed. Why does the test workflow use `contents` btw.?
- [ ] I just saw that `lint-pr-title-preview-outputErrorMessage` mistakenly doesn't use the `.yml` extension. Do you think you could fix that while you're at it?

Thank you also for upgrading some action versions!

Many thanks! 🙌

JoaquinFernandez commented:
Any updates on this? It seems like the right way to permission de action

stevenh commented:
Good to see the addition of permissions, gives users a better understanding of what it needs to do the job. Would love to see this merged.

natterstefan commented:
Hi @stevenh, thanks for the reminder. I keep the tab open now, and I hope to get to this in the next few days.

stevenh commented on README.md:
```diff
@@ -30,6 +30,9 @@ on:
       - edited
       - synchronize
 
+permissions:
```
With this changing to read, can `pull_request_target` above be changed to just `pull_request`?

I ask as a new user I was banging my head trying to figure out why the new workflow didn't trigger. Took longer than I would like to admit to figure out that `pull_request_target` only triggers after it's in base hence it's actually impossible to test during PR submission that it's working as intended.

natterstefan replied:
I do not understand why this change affects the `on` part? Can you explain that further to me, please?

I think, @amannn added `pull_request_target` by design here.

stevenh replied:
I think it was added for security given it needed write so someone could put in a pr and use it to write to the repo before it was approved and merged.

With bit limited to read in your update it would be safe to switch to `pull_request` which allows testing the change prior to merge.

natterstefan replied:
But actually, you can test changes before merging. Check out my referenced PR in my forked repo for this PR: https://github.com/natterstefan/action-semantic-pull-request/pull/1.

You have to create

- one PR for the origin repo
- one PR (only for demo and testing purposes) in your forked repo

That should do the job without changing this.

stevenh replied:
To clarify I meant testing a PR to add this to your own repo for the first time.

Having to merge a PR before you can confirm it works is counter to best practices so having that be the default in the README definitely caught me out, causing much head scratching.

So if the permissions added mean that it can now be safely used for the standard case with just `pull_request` that would be good.

natterstefan replied:
Hi @stevenh,

> Having to merge a PR before you can confirm it works is counter to best practices so having that be the default in the README definitely caught me out, causing much head scratching.

But at the same time, maintainers can be sure that only actions run which were at least added once on the `main` branch, don't they? If so (but I am not 100% sure and not up to date about what is the common way to handle this), I'd leave it as is.

I think this is up to @amannn to decide how to approach your suggestion, but IMHO not in this branch but in another one.

amannn replied:
I'm definitely open to discuss changing the suggested event trigger, but as @natterstefan said, we can do this in a separate PR to not block this one (as it's already been open for a longer time).

natterstefan commented:
Hi @amannn,

> * [ ]  When the `wip` feature is used, we also require to update statuses. Can you add information about this in the comment for the `wip` option? I think generally suggesting read permissions for PRs is a good idea.

> * [ ]  Can you please also link to a test run as it's mentioned in the [contributors guide](https://github.com/amannn/action-semantic-pull-request/blob/main/CONTRIBUTORS.md?rgh-link-date=2022-10-27T06%3A18%3A17Z)? E.g. I'd be curious if using the checkout action only requires the `pull-requests: read` option, or if a permission like `contents` is needed.

I already linked to a Demo PR with test runs in the PR description.

>   Why does the test workflow use `contents` btw.?

I am not 100% sure why I've added `contents: none` in the past but I have now changed it to `contents: read` as this is the only permission this workflow requires in my opinion.

amannn commented:
> I am not 100% sure which information I should add to the [WIP part of the config in README.md](https://github.com/amannn/action-semantic-pull-request/tree/ff373f4e8056b732dfd0eadd42ae54c004e5523b#configuration). Can you explain this further to me, please?

The built-in support for `[WIP]` relies on having write permissions to [change the status of a PR](https://github.com/amannn/action-semantic-pull-request/blob/ff373f4e8056b732dfd0eadd42ae54c004e5523b/src/index.js#L145-L161). If our example configuration advocates for `pull-requests: read`, users of the `wip` option likely have to add another permission—this should be documented.

Does this help?

natterstefan commented:
> The built-in support for `[WIP]` relies on having write permissions to [change the status of a PR](https://github.com/amannn/action-semantic-pull-request/blob/ff373f4e8056b732dfd0eadd42ae54c004e5523b/src/index.js#L145-L161). If our example configuration advocates for `pull-requests: read`, users of the `wip` option likely have to add another permission—this should be documented.
> 
> Does this help?

I got it. It makes sense, and thanks for the info. You see what I came up with for the `README.me` here: [`d1c5b62` (#215)](https://github.com/amannn/action-semantic-pull-request/pull/215/commits/d1c5b6221e1764995cff40578d4a54e8cc88e98c).

Adding a `lint-pr-title-preview-wip.yml` action would be great to test in an example PR (or commit). That allows you to scan the example PR for at least one commit that "tests" the feature. But not sure if that's feasible to ask contributors to add `[WIP]` first and then remove it again after the action has been completed (so you can see it's green) and only then request a review from your side (as they most likely remove it again to proof the other actions work).

amannn commented on README.md:
````diff
@@ -111,6 +114,21 @@ feat(ui): Add `Button` component
           wip: true
 ```
 
+### Required Permissions for [WIP] feature
+
+If you want to use the `[WIP]` feature, you need to grant the
+`pull-requests: write` permission to the GitHub Action. This is because the
+action will update the status of the PR.
+
+```yml
+name: "Lint PR"
+
+# ...
+
+permissions:
+  pull-requests: write
+```
+
````
Looks pretty good!

Maybe we can move the whole config option for `wip` to this block, similar to what we've done with the legacy configuration below? I'm just worried that users stop reading at the `wip` option above and don't notice that permissions need to be changed (even though it's the very next headline).

By doing this, we can make sure we're educating users about the required permissions as we're introducing the feature. Maybe you can reuse the description from above here and adjust the headline accordingly to match the whole section?

I just noticed the "Note that a second check will be reported if this is enabled." part is no longer valid, can you remove this as part of moving the text?

Thank you so much! 🙏

natterstefan replied:
Hey @amannn,

please review again and take a look at:

- [`1ebd749` (#215)](https://github.com/amannn/action-semantic-pull-request/pull/215/commits/1ebd749c2a2e3c7413f0d3e19df1d9d4e6790fa6)
[`f030897` (#215)](https://github.com/amannn/action-semantic-pull-request/pull/215/commits/f030897c39c6656af87f5200b0739d52d7582f75)

amannn replied:
Looks great, thanks! I've included a commit with a few further README improvements that I had on my list to include them in the next release.

amannn reviewed (approved):
Thank you so much @natterstefan! 👏
